src/_pytest/runner.py

```diff
@@ -342,10 +342,26 @@ def from_call(
         instant = timing.Instant()
         try:
             result: TResult | None = func()
-        except BaseException:
+        except BaseException as caught:
             excinfo = ExceptionInfo.from_current()
-            if reraise is not None and isinstance(excinfo.value, reraise):
-                raise
+            val = excinfo.value
+
+            if reraise is not None:
+                reraise_types = (
+                    (reraise,) if not isinstance(reraise, tuple) else reraise
+                )
+
+                # EG-aware path (only direct children; no nesting)
+                if isinstance(val, BaseExceptionGroup):
+                    exception_group = cast(BaseExceptionGroup, val)
+                    for child in exception_group.exceptions:
+                        if isinstance(child, reraise_types):
+                            # B904: chain to the original exception we caught
+                            raise child from caught
+
+                if isinstance(val, reraise_types):  # non-group or no match
+                    raise
```

Comment:
The `cast` is not needed, `isinstance` results in type narrowing.

Some tidying up:

```suggestion
                # ExceptionGroup-aware path: check if any of the direct children
                # is an instance of the `reraise` parameter, and reraise the exception
                # accordingly (#13650).
                if isinstance(val, BaseExceptionGroup):
                    for child in val.exceptions:
                        if isinstance(child, reraise_types):
                            raise child from caught
                # Not an exception group, check if we need to reraise it.            
                elif isinstance(val, reraise_types):  
                    raise
```